ch12.md

```diff
@@ -0,0 +1,256 @@
+const tldr = compose(map(firstWords), readFile)
+
+traverse(Task.of, tldr, files)
+// Task(['hail the monarchy', 'smash the patriarchy'])
+```
+
+Using `traverse` instead of `map`, we've successfully herded those unruly `Task`s into a nice coordinated array of results. This is like `Promise.all()`, if you're familiar, except it isn't just a one-off, custom function, no, this works for any *traversable* type. These mathematical apis tend to capture most things we'd like to do in an interoperable, reusable way, rather than each library reinventing these functions for a single type.
+
+Let's clean up the last example for closure (no, not that kind):
+
+```js
+// getAttribute :: String -> Node -> Maybe String
+// $ :: Selector -> IO Node
+
+// getControlNode :: IO (Maybe Node)
+const getControlNode = compose(chain(traverse(IO.of, $)), map(getAttribute('aria-controls')), $)
+```
+
+Instead of `map(map($))` we have `chain(traverse(IO.of, $))` which inverts our types as it maps then flattens the two `IO`s via `chain`.
+
+## No law and order
+
+Well now, before you get all judgemental and bang the backspace button like a gavel to retreat from the chapter, take a moment to recognize that these laws are useful code guarantees. Tis' my conjecture that the goal of most program architecture is an attempt to place useful restrictions on our code to narrow the possibilities, to guide us into the answers as designers and readers.
+
+An interface without laws is merely indirection. Like any other mathematical structure, we must expose properties for our own sanity. This has a similar effect as encapsulation since it protects the data, enabling us to swap out the interface with another law abiding citizen.
+
+Come along now, we've got some laws to suss out.
+
+### Identity
+
+```js
+const identity1 = compose(sequence(Identity.of), map(Identity.of))
+const identity2 = Identity.of
+
+// test it out with Right
+identity1(Right('stuff'))
+// Identity(Right('stuff'))
+
+identity2(Right('stuff'))
+// Identity(Right('stuff'))
+```
+
+This should be straight forward. If we place an `Identity` in our functor, then turn it inside out with `sequence` that's the same as just placing it on the outside to begin with. We chose `Right` as our guinea pig as it is easy to try the law and inspect. An arbitrary functor there is normal, however, the use of a concrete functor here, namely `Identity` in the law itself might raise some eyebrows. Remember a [category](ch5.md#category-theory) is defined by morphisms between its objects that have associative composition and identity. When dealing with the category of functors, natural transformations are the morphisms and `Identity` is, well identity. The `Identity` functor is as fundamental in demonstrating laws as our `compose` function. In fact, we should give up the ghost and follow suit with our [Compose](ch8.md#a-spot-of-theory) type:
+
+### Composition
+
+```js
+const comp1 = compose(sequence(Compose.of), map(Compose.of))
```

Comment:
Oh gosh, I agree with @KtorZ for dat recap.
Also I know this is a REALLY big can of worms (like how does one implement *really* implement `Compose.of` without a type system), BUT maybe showing `ap` and `Compose.of` implementations (I know that could be its own huge section.). Although it may point out some of the shortcomings that surround traversals in JS and take a lot of focus away from the point here.

That is tricky.

Also should this be:
```js
const comp1 = compose(sequence(Compose.of), map(Compose))
```
wouldn't that `Compose.of` throw in a F(G(F(G(x))))
I did not code it out yet though so I am probably wrong.

Reply:
Also coming with the restructuring. These implementation details are now provided in an appendix. `Compose.of` is a bit tricky to define though in JavaScript. Ideally, we would need to specify the 2 underlying functors' constructors but we can't change the signature of `of` (it will defy its entire purpose, even though being, imo, acceptable for this edge case). The way @DrBoolean solved it in the examples was to directly provide a value nested in 2 functors: 

```js
Compose.of(Identity.of(Either.of(true)));
```

However, we would ideally like to do is something like:

```js
Compose.of(true); // Identity(Right(true))
```

Yet, for the same reasons `sequence` and `traverse` have an extra argument, we can't write this as such in JavaScript like we would [in Haskell](http://hackage.haskell.org/package/base-4.10.0.0/docs/Data-Functor-Compose.html#v:pure). We could go for a "factory-like" method as such:

```js
class Compose {
  static getOf(f, g) {
    return x => new this(f(g(x)));
  }

  constructor(x) {
    this.$value = x;
  }
}

const of = Compose.getOf(Identity.of, Either.of);

of(true); // Compose(Identity(Right(true)))
```

or simply use optional arguments to provide constructors in that case:

```js
class Compose {
  static of(x, f=Identity.of, g=Identity.of) {
    return new Compose(f(g(x)));
  }

  constructor(x) {
    this.$value = x;
  }
}

Compose.of(true, Identity.of, Either.of); // Compose(Identity(Right(true)))
```

Reply:
There's also a closure approach:

```js
const createCompose = (F, G) => {
return class Compose { 
  static of(x) {
    return new Compose(F(G(x)));
  }

  constructor(x) {
    this.$value = x;
  }
}
}
const Compose = createCompose(Id, Either)
Compose.of(true, Identity.of, Either.of); // Compose(Identity(Right(true)))
```

Reply:
I like the closure approach better and will stick with that for the upcoming appendix :+1: